src/client/gl-texture.cc

```diff
@@ -0,0 +1,104 @@
+#include "client/gl-texture.h"
+
+#include "client/atomic-command-queue.h"
+#include "core/logger.h"
+
+namespace zen::remote::client {
+
+GlTexture::GlTexture(uint64_t id, AtomicCommandQueue *update_rendering_queue)
+    : id_(id),
+      update_rendering_queue_(update_rendering_queue),
+      rendering_(new RenderingState())
+{
+}
+
+GlTexture::~GlTexture()
+{
+  auto command = CreateCommand([rendering = rendering_](bool /*cancel*/) {
+    if (rendering->texture_id != 0) {
+      glDeleteTextures(1, &rendering->texture_id);
+    }
+  });
+
+  rendering_.reset();
+
+  update_rendering_queue_->Push(std::move(command));
+}
+
+void
+GlTexture::Commit()
+{
+  if (rendering_->texture_id == 0) {
+    auto command = CreateCommand([rendering = rendering_](bool cancel) {
+      if (cancel) {
+        return;
+      }
+
+      glGenTextures(1, &rendering->texture_id);
+    });
+    update_rendering_queue_->Push(std::move(command));
+  }
```

Comment:
`rendering_` is modified by the rendering thread, so checking value in update thread can cause data race.

```suggestion
  auto command = CreateCommand([rendering = rendering_](bool cancel) {
    if (cancel) {
      return;
    }

    if (rendering->texture_id == 0) {
      glGenTextures(1, &rendering->texture_id);
    }
  });
  update_rendering_queue_->Push(std::move(command));
```